### PR TITLE
add np.stack

### PR DESCRIPTION
**Pull request by sangyx**
I have added the diff function of `np.stack`. 

When we extend the code to higher-order derivatives, the situation becomes a bit more complicated. There are some places to improve:

* In some derivative functions, the program may enter an infinite loop. For example, `np.sum` is used when registering the gradient of `np.sum`. I plan to add depth attributes to `DiffArray` or `to()` to control the order of gradients.
* The function registered in nograd_functions will register the gradient when it is called in the form of `ufunc.__call__`.
* Consider a cleaner way to control whether the calculation should derive.

Do you have any idea about the above questions?

### CODE REVIEW

**Comment by hameerabbasi** [comment]
Ah, yes. I remember running into this when I was initially working with `udiff`. Basically, in your approach, it would work as follows:

- The `DiffArray.to` method should only compute the diff with respect to what's passed in, not anything else.
- `DiffArray.backward` should only compute the calculation graph, but only for "the current level". The next level should remain "unevaluated" or in the form of a graph until you actually call `DiffArray.diff[some_name].backward`.

**Comment by sangyx** [comment]
I have fixed the problem, we have to put all the calculations that need to register gradients in each diff function in `lambda g` (or `vjp`), so as to avoid infinite loops.

The error of `<uarray multimethod '__instancecheck__'>` in ci seem to be related to `uarray`.

**Comment by hameerabbasi** [comment]
I'll fix those up. Don't worry about them for now.

**Comment by hameerabbasi** [comment]
@peterbell10 I can reproduce the segfault in CI. Is it possible for you to investigate? I was also getting a bunch of "returned NULL without setting an error" on `determine_backend`.

**Comment by hameerabbasi** [comment]
@sangyx Passing the torch back to you. 😄 The segfaults  should be resolved now.

Edit: To resolve them locally, get `master` of `uarray` and `unumpy`, and do `pip install -e .` in the directories of both.

**Comment by sangyx** [comment]
Thanks!

**Comment by sangyx** [comment]
@hameerabbasi Hi, I want to add `<uarray multimethod '__instancecheck__'>` to `raw_functions`. But I don't know how to refer it. How can I find it by code like `np.ndim`?

**Comment by hameerabbasi** [comment]
> Hi, I want to add `<uarray multimethod '__instancecheck__'>` to `raw_functions`. But I don't know how to refer it. How can I find it by code like `np.ndim`?

You can probably do something like `unumpy.ClassOverrideMeta.__instancecheck__`.

**Comment by sangyx** [comment]
@hameerabbasi Could you help me to check whether the `NoGradBackend` is correct?

File: src/udiff/_vjp_diffs.py [hunk 1]
```diff
@@ -477,31 +481,54 @@ def grad_transpose(ans, x, axes=None):
...
-        np.where(np.logical_and(np.array(old_shape) == 1, np.array(new_shape) > 1))[0]
-    )
+    from udiff import NoGradBackend
+
+    with ua.set_backend(NoGradBackend(numpy_backend)):
+        broadcast_axes = tuple(
+            np.where(np.logical_and(np.array(old_shape) == 1, np.array(new_shape) > 1))[
+                0
+            ]
+        )
```
Comment (hameerabbasi) [inline src/udiff/_vjp_diffs.py]: Can you write shape manipulations in loops, i.e. completely without NumPy?

File: src/udiff/_uarray_plug.py [hunk 2]
```diff
@@ -1,88 +1,125 @@
...
 
-@wrap_single_convertor
-def __ua_convert__(value, dispatch_type, coerce):
-    if dispatch_type is np.ndarray:
-        if value is None:
-            return value
 
-        if isinstance(value, DiffArray):
-            return value
+class NoGradBackend:
```
Comment (hameerabbasi) [inline src/udiff/_uarray_plug.py]: You can remove `NoGradBackend` and instead use the following directly:

```python
with ua.set_backend(self._inner, coerce=True):
    out = func(*args, **kwargs)
```

File: README.md [hunk 3]
```diff
@@ -8,14 +8,13 @@ import unumpy as np
 import udiff
 from unumpy import numpy_backend
 
-with ua.set_backend(numpy_backend), ua.set_backend(udiff, coerce=True):
+with ua.set_backend(numpy_backend, coerce=True), ua.set_backend(udiff, coerce=True):
```
Comment (hameerabbasi) [inline README.md]: This line needs updating.
```suggestion
with ua.set_backend(udiff.DiffArrayBackend(numpy_backend), coerce=True):
```

Comment (hameerabbasi) [reply README.md]: As do the notebooks.

File: src/tests/test_jacobian.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+#     import xnd
+#     from ndtypes import ndt
+
+#     LIST_BACKENDS.append(XndBackend)
+#     FULLY_TESTED_BACKENDS.append(XndBackend)
+# except ImportError:
+#     XndBackend = None  # type: ignore
+#     LIST_BACKENDS.append(
+#         pytest.param(None, marks=pytest.mark.skip(reason="xnd is not importable"))
+#     )
```
Comment (hameerabbasi) [inline src/tests/test_jacobian.py]: Remove these lines.

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
Thanks, @sangyx. A few minor comments and then it should be good to go.

**Comment by hameerabbasi** [comment]
Thanks @sangyx!

**Pull request merged**